### PR TITLE
Simple build system, UMD ES modules

### LINKED ISSUE #3: Support UMD and ES modules
More and more people are using ES modules because of their static nature. It would be great if Multiavatar supported that. Support for UMD would also be great so then users of AMD and people who just read off the global object could also use it. There a couple different ways this could be implemented, the most organized being a rollup build setup or something, the least being copy and pasting the same code into different files for different formats.

### PR DESCRIPTION
**Pull request by shadowtime2000**
Resolves #3

### CODE REVIEW

File: tools/build.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+        }
+    })(typeof self !== "undefined" ? self : this, function () {
+        ${content};
+        return multiavatar;
+    })
+`
+
+fs.writeFileSync(path.join(__dirname, "../dist/multiavatar.es.js"), `${originalFile.toString()};export default multiavatar;`)
+fs.writeFileSync(path.join(__dirname, "../dist/multiavatar.cjs"), `${originalFile};module.exports=multiavatar;`);
+fs.writeFileSync(path.join(__dirname, "../dist/multiavatar.umd.js"), umd(originalFile));
```
Comment (shadowtime2000) [inline tools/build.js:24]: I feel like this could have been done with a Gulp build system or something instead of having to do some file writes.

**Review by shadowtime2000 — COMMENTED** [review]
To be honest I feel like we could have used some sort of small build system

**Comment by giekaton** [comment]
Thanks for the update. I will review it soon.

Regarding the build system, I'm quite skeptical of frameworks and tools.

They add additional overhead, require updating, sometimes they don't work with some environments, may have strange limitations, or they can be unknown to new users. They also introduce specific coding standards, require additional dependencies that also need updating, and so on.

In my work, I'm following the Occam's razor principle, which states that "the simplest explanation is usually the right one."

Also, when writing everything from scratch, the mind is forced to work based on first principles and then can invent something new.

Of course, when the project is more complex, build systems and frameworks are great, but for Multiavatar I'm intentionally keeping it at the most simple level.

Having a custom-made build system that prepares all three versions seems like an elegant solution. Further on, it can be expanded with additional Multiavatar-specific custom features.

**Comment by giekaton** [comment]
Hey, I have reviewed the code.

So, as I understand, when running `npm publish`, the `prepublishOnly` script will build all required versions, before publishing the package to npm.

The script uses `multiavatar.min.js` which stays the same as before and can be accessed through CDN or manually included in any HTML file. That's good.

Regarding the `del-cli` dependency, I have edited the build script like this `"build": "node tools/build.js"`, and now simply by running `npm run build` or `yarn build`, I get the same result as with the `del-cli`. I guess deleting the `dist` folder is standard practice, but is it necessary in this case as we only have the same three files that are being overwritten with every build?

Also, is there a need to commit `yarn.lock` file? In Yarn docs, it's recommended, but I have also read on StackOverflow that if it's a simple library, the lock file can be ignored. And in general, why did you choose to build with Yarn and not npm?

Regarding minification, you mentioned Terser. The main `multiavatar.js` script also needs to be minified. I'm currently using VS Code extension, which is manual and needs to be automated. Maybe both minification tasks can be integrated, using Terser or a similar tool when making the final build? In this case, the build script would first minify the `multiavatar.js` into `multiavatar.min.js`.

I'm also thinking about rewriting the `svg\_build.php` script into JS. This script extracts the svg code from the `.svg` files, prepares and optimizes it, and then injects it into the dev script `multiavatar.js`. For development needs, it's good to have this as a separate step, not part of the final build script.

**Comment by shadowtime2000** [comment]
> Regarding the del-cli dependency, I have edited the build script like this "build": "node tools/build.js", and now simply by running npm run build or yarn build, I get the same result as with the del-cli. I guess deleting the dist folder is standard practice, but is it necessary in this case as we only have the same three files that are being overwritten with every build?

I was just doing that so if we intend to add or remove module formats or we migrate to a new system or something nothing will be left over in `dist` if we do a `v2` or something.

> Also, is there a need to commit yarn.lock file? In Yarn docs, it's recommended, but I have also read on StackOverflow that if it's a simple library, the lock file can be ignored. And in general, why did you choose to build with Yarn and not npm?

I chose Yarn over `npm` just out of preference because I find Yarn better and/or faster. Yeah, `yarn.lock` doesn't need to be commited but there is no necessarily negative effect of commiting it. 

> Regarding minification, you mentioned Terser. The main multiavatar.js script also needs to be minified. I'm currently using VS Code extension, which is manual and needs to be automated. Maybe both minification tasks can be integrated, using Terser or a similar tool when making the final build? In this case, the build script would first minify the multiavatar.js into multiavatar.min.js.

I think I can implement this, generate `multiavatar.min.js` from `multiavatar.js` and then also minify the CJS, ES, and UMD exportst.

**Comment by giekaton** [comment]
So if `del-cli` is currently not a must-have, let's not use it. We can add it later when needed.

I have also found Yarn to run better than npm in some cases, so let's use Yarn. But if the `yarn.lock` is not necessary, let's ignore it.

Having fewer files/dependencies in the package is easier for new users to understand and start using the code.

For example, if someone who doesn't have/use Node sees `yarn.lock`, it may imply that you must use Node build tools to run the project, while in this case the project can be used even without Node.

If you can add a minifier, as mentioned, that would be great.

**Comment by giekaton** [comment]
- Added some minor fixes for the script, license and package.json
- Removed `del-cli` and `yarn.lock`
- Included contributors section in `package.json`, but you need to update it with your info
- Moved `build_dev.php` to the `tools` folder

Please review and let me know what do you think.

**Comment by giekaton** [comment]
Hi, I have implemented ES module support in a slightly different way. Thanks for the motivation! Closing this PR now.

**Pull request closed without merging**